lib/collection/src/collection/clean.rs

```diff
@@ -0,0 +1,318 @@
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::{Arc, Weak};
+use std::time::{Duration, Instant};
+
+use cancel::{CancellationToken, DropGuard};
+use parking_lot::RwLock;
+use segment::types::{Condition, Filter};
+use tokio::sync::watch::{Receiver, Sender};
+use tokio::task::JoinHandle;
+
+use super::Collection;
+use crate::operations::types::{CollectionError, CollectionResult, UpdateResult, UpdateStatus};
+use crate::operations::{CollectionUpdateOperations, OperationWithClockTag};
+use crate::shards::shard::ShardId;
+use crate::shards::shard_holder::LockedShardHolder;
+use crate::telemetry::ShardCleanStatusTelemetry;
+
+const CLEAN_BATCH_SIZE: usize = 5_000;
+
+/// A collection of local shard clean tasks
+///
+/// Manages tasks for shards and allows easily awaiting a completion state.
+///
+/// These tasks are not persisted in any way and are lost on restart. In case of resharding,
+/// cluster manager will take care of calling the task again and again until it eventually
+/// completes. Once it completes it does not have to be run again on restart.
+#[derive(Default)]
+pub(super) struct ShardCleanTasks {
+    tasks: Arc<RwLock<HashMap<ShardId, ShardCleanTask>>>,
+}
+
+impl ShardCleanTasks {
+    /// Clean a shard and await for the operation to finish
+    ///
+    /// Creates a new task if none is going, the task failed or if the task was invalidated. Joins
+    /// the existing task if one is currently ongoing or if it successfully finished.
+    ///
+    /// If `wait` is `false`, the function will return immediately after creating a new task with
+    /// the `Acknowledged` status. If `wait` is `true` this function will await at most `timeout`
+    /// for the task to finish. If the task is not completed within the timeout the `Acknowledged`
+    /// status is returned.
+    ///
+    /// To probe for completeness this function must be called again repeatedly until it returns
+    /// the `Completed` status.
+    async fn clean_and_await(
+        &self,
+        shards_holder: &Arc<LockedShardHolder>,
+        shard_id: ShardId,
+        wait: bool,
+        timeout: Option<Duration>,
+    ) -> CollectionResult<UpdateStatus> {
+        let mut tasks = self.tasks.upgradable_read();
+
+        // Await existing task if not cancelled or failed
+        if let Some(task) = tasks
+            .get(&shard_id)
+            .filter(|task| !task.is_cancelled_or_failed())
```

Comment:
Are failed tasks eventually pruned from the Map?

What about logging the task error as well?

Reply:
> Are failed tasks eventually pruned from the Map?

Yes. If a task failed it passes this if-block and creates a new task below it. The new task will replace this one, removing the failed task.

If we don't call the clean endpoint again it'll remain until either of:
- we restart Qdrant
- we delete/recreate the collection
- we invalidate this specific shard due to some other reason (like resharding)

> What about logging the task error as well?

Good idea.